### PR TITLE
[PF-1839] cormwell config auto generation by "terra cromwell generate"

### PR DESCRIPTION
**Pull request by ginaygoogle**

### CODE REVIEW

File: src/main/java/bio/terra/cli/command/cromwell/generate.sh [hunk 1]
```diff
@@ -0,0 +1,58 @@
+#!/bin/bash
+
+# Initialize a default Cromwell config. Don't overwrite in case the user has
+# customized their Cromwell config file on their PD.
+cromwell_config=/home/jupyter/cromwell/cromwell.conf
+echo "WORKSPACE_BUCKET: $1"
+echo "GOOGLE_PROJECT: $2"
+if [ ! -f "${cromwell_config}" ]; then
+  cat <<EOF | sudo -E -u jupyter tee "${cromwell_config}"
```
Comment (ginaygoogle) [inline src/main/java/bio/terra/cli/command/cromwell/generate.sh:9]: Hi @yuhuyoyo I followed the AOU https://github.com/all-of-us/workbench/blob/71bd33f976dbe03b2a23fd1c588e826df00b1f26/api/src/main/webapp/static/initialize_notebook_runtime.sh seems like the sudo does not work in server. Do you have idea on it?

Comment (yuhuyoyo) [reply src/main/java/bio/terra/cli/command/cromwell/generate.sh:9]: can you try no sudo? the local process should already be running the script as the user.

also as discussed offline, set cromwell_config as an environment variable that can be specified by the user. Similar to what you did with GOOGLE_PROJECT, WORKSPACE_BUCKET and PET_SA_EMAIL.

File: cromwell.conf [hunk 2]
```diff
@@ -0,0 +1,47 @@
+include required(classpath("application"))
```
Comment (yuhuyoyo) [inline cromwell.conf:1]: delete?

File: src/test/java/unit/CromwellConfig.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+@Tag("unit")
+public class CromwellConfig extends SingleWorkspaceUnit {
+  @Test
+  @DisplayName("app-launch config affects how apps are launched")
+  void cromwellConfig() throws IOException {
+    workspaceCreator.login();
+
+    // `terra workspace set --id=$id`
+    TestCommand.runCommandExpectSuccess("workspace", "set", "--id=" + getUserFacingId());
+
```
Comment (yuhuyoyo) [inline src/test/java/unit/CromwellConfig.java:23]: you might need to create a folder here.

and then at 25, cromwell generate-config --path=/path/to/the/folder/you/just/created.

**Review by yuhuyoyo — COMMENTED** [review]
looks great! just some quick follow-up suggestions from our discussions.

File: src/main/java/bio/terra/cli/command/cromwell/generate.sh [hunk 4]
```diff
@@ -0,0 +1,58 @@
+#!/bin/bash
```
Comment (ginaygoogle) [inline src/main/java/bio/terra/cli/command/cromwell/generate.sh:1]: Not sure the correct place for this script file. I have 3 proposal for this file location. 1 is under the Docker folder, 2 is under tools folder. 3 is here. I choose 3, because learn from WSM also have similar styles put the script file in the same file with the file who called it. (example: https://github.com/DataBiosphere/terra-workspace-manager/tree/main/service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook)

File: src/main/java/bio/terra/cli/command/cromwell/GenerateConfig.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  @Override
+  protected void execute() {
+    String googleProjectId = Context.requireWorkspace().getGoogleProjectId();
+    String petSaEmail = Context.requireUser().getPetSaEmail();
+
+    // Use {WORKSPACE_BUCKET} as a space holder now, until the workspace created with bucket.
+    Context.getConfig()
+        .getCommandRunnerOption()
+        .getRunner()
+        .runToolCommand(
```
Comment (ginaygoogle) [inline src/main/java/bio/terra/cli/command/cromwell/GenerateConfig.java:28]: Follow the `terra app execute` [code](https://github.com/DataBiosphere/terra-cli/blob/main/src/main/java/bio/terra/cli/command/app/Execute.java#L30)the `runToolCommand` does not have return, so it could not deal with the error(like the path is not a real path). Also the `cromwellPath.path` is a string type, not sure if there any better way to require the "real path" input.

Comment (yuhuyoyo) [reply src/main/java/bio/terra/cli/command/cromwell/GenerateConfig.java:28]: i think that's oK. it will still print out error right so the user can re-run the command with a real path.

File: README.md [hunk 6]
```diff
@@ -740,17 +751,19 @@ for more details.
...
-  clone        Clone an existing workspace.
-  create       Create a new workspace.
-  delete       Delete an existing workspace.
-  describe     Describe the workspace.
-  list         List all workspaces the current user can access.
-  list-users   List the users of the workspace.
-  remove-user  Remove a user or group from the workspace.
-  set          Set the workspace to an existing one.
-  update       Update an existing workspace.
+  add-user         Add a user or group to the workspace.
```
[the remark this answers is not in the document]
Comment (yuhuyoyo) [reply README.md:752]: nit: align add a user with above.

File: src/main/java/bio/terra/cli/command/cromwell/generate-config.sh [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    # Disables the Local backend
+    Local.config.root = "/dev/null"
+
+    PAPIv2-beta {
+      actor-factory = "cromwell.backend.google.pipelines.v2beta.PipelinesApiLifecycleActorFactory"
+
+      config {
+        project = "$2"
+        concurrent-job-limit = 10
+        root = "$3/workflows/cromwell-executions"
```
Comment (yuhuyoyo) [inline src/main/java/bio/terra/cli/command/cromwell/generate-config.sh:33]: let's just hard code it here since we are not passing in a real bucket path. 

but also add a comment saying: replace {WORKSPACE_BUCKET} with the path to a gcs bucket in this workspace, e.g. gs://my-cromwell-workflow-bucket/workflows/cromwell-executions.

File: src/main/java/bio/terra/cli/command/cromwell/generate-config.sh [hunk 8]
```diff
@@ -0,0 +1,58 @@
+#!/bin/bash
+
+# Initialize a default Cromwell config. Don't overwrite in case the user has
+# customized their Cromwell config file on their PD.
+echo "cromwell_config" : $1
```
Comment (yuhuyoyo) [inline src/main/java/bio/terra/cli/command/cromwell/generate-config.sh:5]: CROMWELL_CONFIG_PATH

File: README.md [hunk 9]
```diff
@@ -471,6 +469,17 @@ Currently the available configuration properties are:
...
+#### Cromwell
+
+This command is creating an auto generation cromwell.conf in workspace.
+
+```
+Usage: terra cromwell [COMMAND]
+Generate a cromwell config in workspace ai notebook.
+Commands:
+  generate-config  Autogenerate a cromwell.conf under the user specified path.
+```
```
Comment (yuhuyoyo) [inline README.md:481]: Please also add that user needs to update the root to point to a gcs bucket of their choice.

### Cromwell with jupyter notebook
Cromwell and [cromshell](https://github.com/broadinstitute/cromshell) are installed on notebook under /home/jupyter/cromwell directory.  First, find your cromwell.conf under /home/jupyter/cromwell and update the gcs bucekt. 

Then start the cromwell server:

```
java -Dconfig.file=path/to/cromwell.conf -jar cromwell-81.jar server
```

Cromwell server will be running on localhost:8000.

Open another terminal window and type `cromshell`. On first time usage, it will prompt you to enter the port where cromwell server is running. Set the server to `localhost:8000` in cromshell. Then you can start running your workflow using cromshell in the notebook. e.g. ` cromshell submit workflow.wdl inputs.json options.json dependencies.zip`.  See https://github.com/broadinstitute/cromshell for details

something like this. You can rephrase it to make it clearer.

File: README.md [hunk 10]
```diff
@@ -317,17 +317,14 @@ Commands:
   auth       Retrieve and manage user credentials.
   bq         Call bq in the Terra workspace.
   config     Configure the CLI.
+  cromwell   Generate a cromwell config in workspace ai notebook.
```
Comment (melissachang) [inline README.md:320]:
```
cromwell Generate a Cromwell configuration file.
```

File: src/main/java/bio/terra/cli/command/cromwell/GenerateConfig.java [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+import bio.terra.cli.command.shared.options.CromwellPath;
+import java.util.List;
+import java.util.Optional;
+import picocli.CommandLine;
+import picocli.CommandLine.Command;
+
+/** This class corresponds to the third-level "terra cromwell generate-config" command. */
+@Command(
+    name = "generate-config",
+    description = "Autogenerate a cromwell.conf under the user specified path.")
```
Comment (melissachang) [inline src/main/java/bio/terra/cli/command/cromwell/GenerateConfig.java:14]:
```
description = "Generate cromwell.conf under the user-specified path.")
```

File: src/main/java/bio/terra/cli/command/Cromwell.java [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+import bio.terra.cli.command.cromwell.GenerateConfig;
+import picocli.CommandLine.Command;
+
+/**
+ * This class corresponds to the second-level "terra cromwell" command. This command is not valid by
+ * itself; it is just a grouping keyword for it sub-commands.
+ */
+@Command(
+    name = "cromwell",
+    description = "Generate a cromwell config in workspace ai notebook.",
```
Comment (melissachang) [inline src/main/java/bio/terra/cli/command/Cromwell.java:12]:
```
description = "Commands related to Cromwell workflows.",
```

File: src/main/java/bio/terra/cli/command/shared/options/CromwellPath.java [hunk 13]
```diff
@@ -0,0 +1,12 @@
+package bio.terra.cli.command.shared.options;
```
Comment (melissachang) [inline src/main/java/bio/terra/cli/command/shared/options/CromwellPath.java:1]: `--path` sounds generic. We may want to use for another command in the future. Instead of defining `shared/options/CromwellPath.java`, can you define this option in GenerateConfig.java ?

You require that CromwellPath ends in `cromwell.config`, but you don't validate it. How about passing in directory instead of file path?

```
  @CommandLine.Option(
      names = "--dir",
      defaultValue = ".",
      required = false,
      description =
          "Directory to put generated cromwell.conf in. Defaults to current directory.")
  public String dir;
```

I think it's cleaner to put default in the CommandLine.Option, rather than in code.

File: src/test/java/unit/CromwellConfig.java [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+/** Tests for the `terra cromwell` commands. */
+@Tag("unit")
+public class CromwellConfig extends SingleWorkspaceUnit {
+  @Test
+  @DisplayName("cromwell config create in root path")
+  void cromwellConfigCreateRoot() throws IOException {
+    TestUser testUser = TestUser.chooseTestUserWithSpendAccess();
+    testUser.login();
+    WorkspaceUtils.createWorkspace(testUser);
```
Comment (melissachang) [inline src/test/java/unit/CromwellConfig.java:23]: You don't need to create a workspace. `SingleWorkspaceUnit` creates a workspace. It's called SingleWorkspaceUnit because it creates a single workspace, and it's used for unit tests.

File: src/test/java/unit/CromwellConfig.java [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+import java.io.IOException;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Tag;
+import org.junit.jupiter.api.Test;
+
+/** Tests for the `terra cromwell` commands. */
+@Tag("unit")
+public class CromwellConfig extends SingleWorkspaceUnit {
+  @Test
+  @DisplayName("cromwell config create in root path")
```
Comment (melissachang) [inline src/test/java/unit/CromwellConfig.java:19]: Let's have documentation/test name match actual command.

```
  @DisplayName("cromwell generate-config with default dir")
  void cromwellGenerateConfig() throws IOException {
```

File: src/test/java/unit/CromwellConfig.java [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+
+    // Remove the created config file.
+    new File("cromwell.conf").delete();
+
+    // `terra workspace delete`
+    TestCommand.runCommandExpectSuccess("workspace", "delete", "--quiet");
+  }
+
+  @Test
+  @DisplayName("cromwell config create in test path")
```
Comment (melissachang) [inline src/test/java/unit/CromwellConfig.java:39]:
```
  @DisplayName("cromwell generate-config with custom dir")
  void cromwellGenerateConfigCustomDir() throws IOException {
``

File: src/main/java/bio/terra/cli/command/cromwell/GenerateConfig.java [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+    // Use {WORKSPACE_BUCKET} as a space holder now, until the workspace created with bucket.
+    Context.getConfig()
+        .getCommandRunnerOption()
+        .getRunner()
+        .runToolCommand(
+            List.of(
+                "src/main/java/bio/terra/cli/command/cromwell/generate-config.sh",
+                Optional.ofNullable(cromwellPath.path).orElse("cromwell.conf"),
+                googleProjectId,
+                "{WORKSPACE_BUCKET}",
```
Comment (melissachang) [inline src/main/java/bio/terra/cli/command/cromwell/GenerateConfig.java:33]: Outputing `{WORKPSACE_BUCKET}` isn't useful:

```
 COMMAND: cromwell generate-config --path=build/cromwell.conf
    STDOUT --------------
    Setting the gcloud project to the workspace project
    Updated property [core/project].

    cromwell_config : build/cromwell.conf
    GOOGLE_PROJECT: terra-wsm-d-powerful-lime-7770
    WORKSPACE_BUCKET: {WORKSPACE_BUCKET}
    PET_SA_EMAIL: pet-106477532013210995066@terra-wsm-d-powerful-lime-7770.iam.gserviceaccount.com
```

Let's instead print:

```
IMPORTANT: In {cromwell.conf path}, change `{WORKSPACE_BUCKET}` to a bucket in your workspace.
```

**Review by yuhuyoyo — APPROVED** [review]

File: README.md [hunk 18]
```diff
@@ -471,6 +469,27 @@ Currently the available configuration properties are:
...
+  generate-config  Generate cromwell.conf under the user-specified path.
+```
+
+To run Cromwell in a notebook instance:
+
+* Run `terra cromwell generate-config [--dir=my/path]`
+* In a notebook instance terminal, in `cromwell.conf`, replace `{WORKSPACE_BUCKET}` with a bucket in your workspace
+* Run `java -Dconfig.file=path/to/cromwell.conf -jar cromwell-81.jar server`. This starts Cromwell server on `localhost:8000`.
+* In another terminal window, run `cromshell`. Enter `localhost:8000` for cromwell server.
+* Run workflow: `cromshell submit workflow.wdl inputs.json options.json dependencies.zip`
```
Comment (yuhuyoyo) [inline README.md:489]: Start workflow through cromshell: e.g. `cromshell submit workflow.wdl inputs.json options.json dependencies.zip`

**Review by yuhuyoyo — APPROVED** [review]

File: README.md [hunk 19]
```diff
@@ -471,6 +469,17 @@ Currently the available configuration properties are:
 [format] output format = TEXT
 ```
 
+#### Cromwell
```
[the remark this answers is not in the document]
Comment (melissachang) [reply README.md:472]: N/m, I was at an old commit

File: README.md [hunk 20]
```diff
@@ -471,6 +469,27 @@ Currently the available configuration properties are:
...
+Commands related to Cromwell workflows.
+Commands:
+  generate-config  Generate cromwell.conf under the user-specified path.
+```
+
+To run Cromwell in a notebook instance:
+
+* Run `terra cromwell generate-config [--dir=my/path]`
+* In a notebook instance terminal, in `cromwell.conf`, replace `{WORKSPACE_BUCKET}` with a bucket in your workspace
+* Run `java -Dconfig.file=path/to/cromwell.conf -jar cromwell-81.jar server`. This starts Cromwell server on `localhost:8000`.
```
Comment (melissachang) [inline README.md:487]: cromwell-81.jar -> cromwell/cromwell-81.jar

File: README.md (hunk 18 above)
Comment (melissachang) [reply README.md:489]: Start workflow through cromshell: e.g. `cromshell submit workflow.wdl inputs.json [options.json] [dependencies.zip]`

File: src/main/java/bio/terra/cli/command/cromwell/generate-config.sh [hunk 21]
```diff
@@ -0,0 +1,59 @@
+#!/bin/bash
+
+# Initialize a default Cromwell config. Don't overwrite in case the user has
+# customized their Cromwell config file on their PD.
+echo "CROMWELL_CONFIG_PATH" : $1
+echo "GOOGLE_PROJECT": $2
+echo "PET_SA_EMAIL": $3
+echo "IMPORTANT: In {cromwell.conf path}, change {WORKSPACE_BUCKET} to a bucket in your workspace."
```
Comment (melissachang) [inline src/main/java/bio/terra/cli/command/cromwell/generate-config.sh:8]: {cromwell.conf path} -> cromwell.conf

**Review by melissachang — APPROVED** [review]
Cool
